cli/src/config.rs

```diff
@@ -162,27 +163,65 @@ impl AsMut<StackedConfig> for RawConfig {
     }
 }
 
+#[derive(Copy, Clone, Debug)]
+struct DeprecatedLocation {
+    old: &'static str,
+    new: &'static str,
+}
+#[derive(Clone, Debug)]
+enum ConfigPathState {
+    New,
+    Exists,
+    Deprecated(DeprecatedLocation),
+}
+
+/// A ConfigPath can be in one of two states:
+///
+/// - exists(): a config file exists at the path
+/// - !exists(): a config file doesn't exist here, but a new file _can_ be
+///   created at this path
+///
+/// Additionally, if it exists, it may be deprecated. In this case, jj will warn
+/// when ConfigEnv::warn_for_deprecated_paths is called.
 #[derive(Clone, Debug)]
-enum ConfigPath {
-    /// Existing config file path.
-    Existing(PathBuf),
-    /// Could not find any config file, but a new file can be created at the
-    /// specified location.
-    New(PathBuf),
+pub struct ConfigPath {
+    path: PathBuf,
+    state: ConfigPathState,
 }
 
 impl ConfigPath {
     fn new(path: PathBuf) -> Self {
-        if path.exists() {
-            ConfigPath::Existing(path)
-        } else {
-            ConfigPath::New(path)
+        use ConfigPathState::*;
+        ConfigPath {
+            state: if path.exists() { Exists } else { New },
+            path,
         }
     }
 
-    fn as_path(&self) -> &Path {
-        match self {
-            ConfigPath::Existing(path) | ConfigPath::New(path) => path,
+    fn new_deprecated(path: PathBuf, old: &'static str, new: &'static str) -> Self {
+        use ConfigPathState::*;
+        let state = if path.exists() {
+            Deprecated(DeprecatedLocation { old, new })
+        } else {
+            New
+        };
+
+        ConfigPath { state, path }
```

Comment:
nit: same thing here

```suggestion
        ConfigPath { path, state }
```